x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvMedianAbsoluteDeviation.java

```diff
@@ -0,0 +1,357 @@
+        return mad;
+    }
+
+    /**
+     * If the values are ascending pick the middle value or average the two middle values together.
+     */
+    static long ascending(Longs longs, LongBlock values, int firstValue, int count) {
+        if (longs.values.length < count) {
+            longs.values = ArrayUtil.grow(longs.values, count);
+        }
+        int middle = firstValue + count / 2;
+        long median = count % 2 == 1 ? values.getLong(middle) : avgWithoutOverflow(values.getLong(middle - 1), values.getLong(middle));
+        for (int i = 0; i < count; i++) {
+            long value = values.getLong(firstValue + i);
+            // From here, this array contains unsigned longs
+            longs.values[i] = unsignedDifference(value, median);
+        }
+        return NumericUtils.unsignedLongAsLongExact(unsignedLongMedianOf(longs.values, count));
+    }
+
+    static long single(long value) {
+        return 0L;
+    }
+
+    static long longMedianOf(long[] values, int count) {
+        // TODO quickselect
+        Arrays.sort(values, 0, count);
+        int middle = count / 2;
+        return count % 2 == 1 ? values[middle] : avgWithoutOverflow(values[middle - 1], values[middle]);
+    }
+
+    static class Doubles {
+        public double[] values = new double[2];
+        public int count;
+    }
+
+    @MvEvaluator(extraName = "Double", finish = "finish", single = "single")
+    static void process(Doubles doubles, double v) {
+        if (doubles.values.length < doubles.count + 1) {
+            doubles.values = ArrayUtil.grow(doubles.values, doubles.count + 1);
+        }
+        doubles.values[doubles.count++] = v;
+    }
+
+    static double finish(Doubles doubles) {
+        double median = doubleMedianOf(doubles.values, doubles.count);
+        for (int i = 0; i < doubles.count; i++) {
+            double value = doubles.values[i];
+            // Double differences between median and the values may potentially result in +/-Infinity.
+            // As we use that value just to sort, the MAD should remain finite.
+            doubles.values[i] = value > median ? value - median : median - value;
+        }
+        double mad = doubleMedianOf(doubles.values, doubles.count);
+        doubles.count = 0;
+        return mad;
+    }
+
+    static double ascending(Doubles doubles, DoubleBlock values, int firstValue, int count) {
+        if (doubles.values.length < count) {
+            doubles.values = ArrayUtil.grow(doubles.values, count);
+        }
+        int middle = firstValue + count / 2;
+        double median = count % 2 == 1 ? values.getDouble(middle) : (values.getDouble(middle - 1) + values.getDouble(middle)) / 2;
+        for (int i = 0; i < count; i++) {
+            double value = values.getDouble(firstValue + i);
+            // Double differences between median and the values may potentially result in +/-Infinity.
+            // As we use that value just to sort, the MAD should remain finite.
```

Comment:
The first median shouldn't be infinite, as all values used to calculate it are finite.

For the second median, the median of differences, I _believe_ **if** it can happen that the median is non-finite, we still do a throw in the `doubleMedianOf()` function used later. For all types, we do that check there

Reply:
PS: While doing a change for another comment, I saw that I forgot the `ascending="ascending"` for doubles. And it had an outdated avg calculation `(a + b) / 2` instead of `a / 2 + b / 2`. Just fixed it

Reply:
:+1: